test/distributed/test_c10d.py

```diff
@@ -2989,6 +2997,165 @@ def test_param_layout_mismatch_error(self):
+        process_group = c10d.ProcessGroupGloo(store, self.rank, self.world_size)
+
+        model = DistributedDataParallel(
+            TestDdpCommHook(),
+            process_group=process_group
+        )
+
+        with self.assertRaisesRegex(TypeError, 'Communication hook must be callable.'):
+            model._register_comm_hook(state=None, hook=1)
+
+        with self.assertRaisesRegex(ValueError, 'bucket annotation is not dist.GradBucket.'):
+            def comm_hook(state: object, bucket: int) -> torch.futures.Future:
+                return torch.futures.Future()
+
+            model._register_comm_hook(state=None, hook=comm_hook)
+
+    @requires_gloo()
+    def test_ddp_invalid_comm_hook_return_type(self):
+        """
+        This test checks whether return annotation checked properly if defined. It also
+        checks whether an internal error is thrown if return type is incorrect and user
+        hasn't specified any return type annotation.
+        """
+        store = c10d.FileStore(self.file_name, self.world_size)
+        process_group = c10d.ProcessGroupGloo(store, self.rank, self.world_size)
+
+        model = DistributedDataParallel(
+            TestDdpCommHook(),
+            process_group=process_group
+        )
+
+        with self.assertRaisesRegex(ValueError, 'return annotation is not torch.futures.Future.'):
+            def comm_hook(state: object, bucket: dist.GradBucket) -> int:
+                return torch.futures.Future()
+
+            model._register_comm_hook(state=None, hook=comm_hook)
+
+        with self.assertRaisesRegex(RuntimeError, 'callback must return a torch.futures.Future object, but got'):
+            def comm_hook(state: object, bucket: dist.GradBucket):
+                return 1
+
+            model._register_comm_hook(state=None, hook=comm_hook)
+
+            # Run forward
+            output = model(8, self.rank)
+
+            # Run backward
+            output.mean().backward()
+
+    @requires_gloo()
+    def test_ddp_comm_hook_register_just_once(self):
+        """
+        DDP communication hook can only be registered once. This test validates whether
+        the error is thrown properly when register_comm_hook is called more than once.
+        """
+        store = c10d.FileStore(self.file_name, self.world_size)
+        process_group = c10d.ProcessGroupGloo(store, self.rank, self.world_size)
+
+        model = DistributedDataParallel(
+            TestDdpCommHook(),
+            process_group=process_group
+        )
+
+        def dummy_hook(state, bucket):
+            fut = torch.futures.Future()
+            fut.set_result(bucket.get_tensors())
+            return fut
+
+        dist.PythonHookBinder.register_comm_hook(model, None, dummy_hook)
+
+        with self.assertRaisesRegex(RuntimeError, "register_comm_hook can only be called once."):
+            dist.PythonHookBinder.register_comm_hook(model, None, dummy_hook)
```

Comment:
I feel we shouldn't expose a public API like this, the user should always call `ddp.register_comm_hook`.